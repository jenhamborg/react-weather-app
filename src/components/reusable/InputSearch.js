import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Input = styled.div`
  align-items: center;
  margin-top: 0.5rem;
  position: relative;
  padding: 1rem;
  .input-container {
    position: relative;
  }
  .input-error-message {
    font-size: 1.25rem;
  }
  input {
    color: #f8f8f8;
    padding: 0.7rem 1.25rem;
    font-size: 1.35rem;
    border-radius: 4px;
    border: 1px solid rgba(255, 255, 255, 0.8);
    background: rgba(0, 0, 0, 0.1);
    height: 50px;
    ::placeholder {
      color: #f8f8f8;
      font-size: 1.25rem;
    }
  }
`;

function InputSearch({
  errorMessage,
  keyFunction,
  label,
  maxCharacter,
  placeholder,
  userValue,
  userValueSetter,
  type,
}) {
  return (
    <Input>
      <div className="input-container">
        <input
          data-testid="input"
          aria-labelledby={label}
          type={type}
          label={label}
          maxLength={maxCharacter}
          onKeyPress={keyFunction}
          onChange={(e) => userValueSetter(e.target.value)}
          value={userValue}
          placeholder={placeholder}
        />
      </div>
      <div className="input-error-message">{errorMessage && errorMessage}</div>
    </Input>
  );
}

export default InputSearch;

InputSearch.propTypes = {
  errorMessage: PropTypes.string,
  label: PropTypes.string,
  keyFunction: PropTypes.func,
  maxCharacter: PropTypes.string,
  placeholder: PropTypes.string,
  setErrorMessage: PropTypes.func,
  userValue: PropTypes.string,
  userValueSetter: PropTypes.func,
  userInput: PropTypes.func,
  type: PropTypes.string,
};
