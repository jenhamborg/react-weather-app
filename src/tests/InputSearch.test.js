import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import InputSearch from "../components/reusable/searchInput";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<InputSearch />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Input Search Component", () => {
  const mockFunction = jest.fn();
  afterEach(() => {
    cleanup;
  });

  it("function is called with input value onKeyPress", async () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation((init) => [init, setState]);

    const { getByTestId } = render(
      <InputSearch
        data-testid="input"
        errorMessage={"Error!"}
        keyFunction={mockFunction}
        label={"enter city or zip code and hit enter"}
        maxCharacter={"10"}
        placeholder={"Enter City or Zip Code"}
        userValue={"test"}
        userValueSetter={setState}
        type={"search"}
      />
    );

    const inputNode = getByTestId("input");
    act(() => {
      fireEvent.change(inputNode, { target: { value: "Minneapolis" } });
      fireEvent.keyPress(inputNode, { key: "Enter", code: 13, charCode: 13 });
    });
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  it("onChange hits mock setter with expected data", async () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation((init) => [init, setState]);

    const { getByTestId } = render(
      <InputSearch
        data-testid="input"
        errorMessage={"Error!"}
        keyFunction={mockFunction}
        label={"enter city or zip code and hit enter"}
        maxCharacter={"10"}
        placeholder={"Enter City or Zip Code"}
        userValue={"test"}
        userValueSetter={setState}
        type={"search"}
      />
    );

    const inputNode = getByTestId("input");

    fireEvent.change(inputNode, { target: { value: "Minneapolis" } });

    expect(setState).toHaveBeenCalledWith("Minneapolis");
  });
});
