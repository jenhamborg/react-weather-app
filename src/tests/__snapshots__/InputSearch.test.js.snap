// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="searchInput__Input-sc-7ewft-0 kzurIl"
>
  <div
    className="input-container"
  >
    <input
      data-testid="input"
      onChange={[Function]}
    />
  </div>
  <div
    className="input-error-message"
  />
</div>
`;
